src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs

```diff
@@ -2,108 +2,1037 @@
+            /// </summary>
+            private MessageHeader _lastReceivedHeader;
+            /// <summary>Temporary buffer used to store the header for received messages.</summary>
+            private byte[] _receiveHeaderBuffer = new byte[MaxMessageHeaderLength];
+            /// <summary>
+            /// Buffer used to store the complete message to be sent to the stream.  This is needed
+            /// rather than just sending a header and then the user's buffer, as we need to mutate the
+            /// buffered data with the mask, and we don't want to change the data in the user's buffer.
+            /// </summary>
+            private byte[] _sendBuffer;
+            /// <summary>
+            /// Whether the last SendAsync had endOfMessage==false. We need to track this so that we
+            /// can send the subsequent message with a continuation opcode if the last message was a fragment.
+            /// </summary>
+            private bool _lastSendWasFragment;
+
+            // Thread-safety:
+            // It's acceptable to call ReceiveAsync and SendAsync in parallel.  One of each may run concurrently.
+            // Attemping to invoke any other operations in parallel may corrupt the instance.  Attempting to invoke
+            // a send operation while another is in progress or a receive operation while another is in progress will
+            // result in an exception.
+
+            /// <summary>
+            /// The task returned from the last SendAsync operation to not complete synchronously.
+            /// If this is not null and not completed when a subsequent SendAsync is issued, an exception occurs.
+            /// </summary>
+            private Task _lastSendAsync;
+            /// <summary>
+            /// The task returned from the last ReceiveAsync operation to not complete synchronously.
+            /// If this is not null and not completed when a subsequent ReceiveAsync is issued, an exception occurs.
+            /// </summary>
+            private Task<WebSocketReceiveResult> _lastReceiveAsync;
+            /// <summary>
+            /// Semaphore used to ensure that calls to SendFrameAsync don't run concurrently.  While <see cref="_lastSendAsync"/>
+            /// is used to fail if a caller tries to issue another SendAsync while a previous one is running, internally
+            /// we use SendFrameAsync as an implementation detail, and it should not cause user requests to SendAsync to fail,
+            /// nor should such internal usage be allowed to run concurrently with other internal usage or with SendAsync.
+            /// </summary>
+            private readonly SemaphoreSlim _sendFrameAsyncLock = new SemaphoreSlim(1, 1);
+
+            public ManagedClientWebSocket()
             {
-                throw new PlatformNotSupportedException(SR.net_WebSockets_UnsupportedPlatform);
+                // Initialize the keep alive timer.  When active, it'll send periodic ping frames to the server.
```

Comment:
How critical is the timing of this?  What if a ping is delayed for a very long time?  Given that Timers might be delayed by other items in the ThreadPool, should we consider a more dependable time source?